### PR TITLE
[fern] Round-1: Gaussian random Fourier coord features + 4L/512d/8h

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The Transolver backbone receives raw `(x, y, z)` coordinates plus normals
and a single SDF value as features. Coordinate inputs are spectrally
biased — early MLPs learn smooth coordinate functions slowly because all
spatial scales are encoded in the same low-frequency basis. **Gaussian
random Fourier features (RFF)** apply a fixed random projection
`gamma * 2π * (x · B)` (with `B` Gaussian-distributed) before the trig
transform, lifting the input into a high-frequency basis where the model
can fit fine geometric details quickly. RFF is well-established in
implicit neural representations (NeRF, Tancik et al. 2020) and was the
proven coordinate-encoding winner on `radford` (per yi research notes).

On DrivAerML this should help most where the model needs to capture
fine-scale surface variation (`surface_pressure`, wall shear) — i.e. the
exact axes furthest from AB-UPT.

## Instructions

Add Gaussian RFF coord encoding for both surface and volume inputs at
fixed dimensionality. Compose with the 4L/512d/8h Transolver config.

### Implementation

In `model.py` or a new `model_rff.py`, add:

```python
class RFFEncoding(nn.Module):
    def __init__(self, in_dim: int, num_features: int = 32, sigma: float = 1.0):
        super().__init__()
        # Fixed random projection — not trainable
        self.register_buffer("B", torch.randn(in_dim, num_features) * sigma)
    def forward(self, x):
        # x: [..., in_dim] -> [..., 2 * num_features]
        proj = 2 * math.pi * (x @ self.B)
        return torch.cat([proj.sin(), proj.cos()], dim=-1)
```

Apply RFF only to the **coordinate channels** of surface and volume
inputs:
- Surface input: `[x, y, z, nx, ny, nz, area]` (channels 0-2 are coords).
- Volume input: `[x, y, z, sdf]` (channels 0-2 are coords).

Concatenate `RFF(coords)` with the original feature tensor (so the model
keeps access to raw coords, normals, area, sdf).

Wire two CLI flags:
- `--rff-num-features 32` (per-axis count → output is 64 per axis = 192
  total before sin/cos = 384 after).
- `--rff-sigma 1.0` (controls bandwidth).

**Sweep `--rff-sigma` ∈ {0.5, 1.0, 2.0}** in a 3-arm wandb group. Sigma
controls the bandwidth — too low and RFF gives no benefit over identity;
too high and the model can't converge because every coordinate looks
random. Tancik 2020 found `sigma ≈ 1.0` works well at meter-scale
coordinates.

### Reproduce (single arm, sigma=1.0)

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --wandb-name "fern/rff-coord-sigma1.0" \
  --wandb-group "fern-rff-sigma-sweep" \
  --volume-loss-weight 2.0 \
  --batch-size 4 \
  --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --rff-num-features 32 --rff-sigma 1.0 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

Then submit two more arms with `--rff-sigma 0.5` and `--rff-sigma 2.0`.
If 8-GPU DDP fits all three sequentially in the time budget, run all
three; otherwise pick `sigma=1.0` plus the closer of `0.5`/`2.0` based on
loss curve diagnostics.

### Diagnostics

- Compare `val_primary/surface_pressure_rel_l2_pct` per epoch — RFF should
  lift this faster than the no-RFF comparator (alphonse's calibration).
- Watch param count delta — RFF buffer is non-trainable, so trainable
  count should be unchanged. Total tensor count grows.
- Confirm coords are properly normalized **before** RFF (else `sigma`
  interpretation depends on case-scale variance).

## Reporting

- W&B group link covering all arms run.
- Per-axis `test_primary/*` table for the best arm vs alphonse calibration.
- Comment on which axis benefited most (predict `surface_pressure` and
  `wall_shear`).
- Param count delta + wall time delta vs no-RFF.

## Baseline

Tay alphonse calibration is the live comparator. Reference yi merged
config (no RFF): `abupt_axis_mean = 16.64`.

| Target | This-repo metric | AB-UPT |
|---|---|---:|
| `surface_pressure_rel_l2_pct` | `test_primary/surface_pressure_rel_l2_pct` | **3.82** |
| `wall_shear_rel_l2_pct` | `test_primary/wall_shear_rel_l2_pct` | **7.29** |
| `volume_pressure_rel_l2_pct` | `test_primary/volume_pressure_rel_l2_pct` | **6.08** |

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT fern: heads-up on compile bug

Killed and re-launched sigma=1.0 with `--no-compile-model` after seeing alphonse's report on PR #30 — same DDP partial-last-batch torch.compile assertion would have killed this run at step 2720/2721. The first attempt got 35% into epoch 1 (~13 GPU-min wasted) before I caught it. Re-launched at 10:44 UTC with the fix; I'll apply `--no-compile-model` to the sigma=0.5 and sigma=2.0 arms when they go.

Per-step throughput will drop ~1.5-2x vs compiled (per alphonse), so I expect ~10-12 epochs in the 6h timeout instead of ~13-14. Still enough for clear curves and final test eval.

**Comment by morganmcg1** [comment]
STUDENT fern: implementation review + sweep orchestration

Picked up the in-progress sigma=1.0 run (W&B `u43lik5d`, started 10:44 UTC after the `--no-compile-model` fix). Reviewed the existing implementation and confirmed:

- `RFFEncoding` (`model.py:49`) registers `B` as a non-trainable buffer with `torch.randn(in_dim, num_features) * sigma`; forward applies `2π·(x@B)` then `[sin, cos]` → `2*num_features` channels.
- Surface and volume each get their own RFF instance (different `B`).
- RFF output is concatenated with the original extra features (normals, area, sdf) and fed into the existing `project_*_features` Linear which projects to `n_hidden` and adds to the existing `pos_embed` (so RFF augments rather than replaces sincos PE).
- Trainable param delta vs no-RFF: the projection input grows from 4 → 4+64=68 channels per modality, adding `64 × 512 × 2 = 65,536` trainable params (12.77M total at this config).

**Current sigma=1.0 run health (W&B `u43lik5d`, step 1021/2721 of epoch 1):**
- `train/base_mse_loss`: 0.481 (down from initial; consistent decline)
- `train/grad/global_norm` clipping at 1.0; `global_norm_pre_clip` ≈ 14.6 → clipping is engaging hard (expected at this LR)
- No NaN/Inf in grads; throughput 1.49 it/s (matches advisor's compile-off estimate)
- All 8 GPUs ~71GB / 96GB, 92-100% util

**Coordinate-scale note (worth flagging):** Surface coords are ~meter-scale ([-1, 4]m typical), so `sigma=1.0` matches the Tancik 2020 recommendation. Volume coords are wider (~[-40, 80]m far-field). At `sigma=1.0` the volume RFF projection magnitude saturates well past π for far-field samples, which means those buffer outputs are effectively uncorrelated noise. The existing `pos_embed` (multi-wavelength sincos) covers this range, so the model can route around any noisy RFF channels via the projection — but the `--rff-sigma` sweep is the right way to find the best bandwidth here. Not modifying the implementation mid-sweep; calling this out for follow-up.

**Sweep orchestration:** Launched a detached `run_sweep.sh` (pid 20078) that waits for the current sigma=1.0 to exit, then runs `sigma=0.5` then `sigma=2.0` sequentially with `--no-compile-model`. Each arm uses all 8 GPUs; ~6h timeout each → all 3 arms complete in ~18h wall time. Logs at `target/outputs/fern-rff/{sigma1.0_v2,sigma0.5,sigma2.0}.log`, orchestrator log at `sweep_orchestrator.log`.

Will report results once all three finish, with per-axis test metrics, sigma comparison, and recommendation.

**Comment by morganmcg1** [comment]
ADVISOR: mark ready now — sigma=1.0 is the clear winner, kill the other arms

Your sigma=1.0 primary arm finished with **test_abupt = 17.774** (W&B `u43lik5d`, val 17.06), beating baseline 19.81 by **−2.04 abupt (−10.3%)**. This is the new tay leader. Full per-axis breakdown:

| Metric | fern test | tay baseline | yi best | AB-UPT |
|---|---:|---:|---:|---:|
| `abupt_axis_mean` | **17.77** | 19.81 | 15.82 | — |
| `surface_pressure` | **11.20** | 12.86 | 9.99 | 3.82 |
| `wall_shear` | **18.68** | 21.27 | 16.60 | 7.29 |
| `volume_pressure` | **16.13** | 15.91 | 14.21 | 6.08 |
| `tau_x` | **16.20** | 18.24 | 14.27 | 5.35 |
| `tau_y` | **21.81** | 25.50 | 19.49 | 3.65 |
| `tau_z` | **23.54** | 26.53 | 21.12 | 3.63 |

RFF lifts every axis except volume_pressure (parity). tau axes drop ~9-13%, surface_pressure ~13%. This is a real win.

**Action items:**

1. **Kill the sigma=0.5 arm immediately.** It's at val_abupt=74.31 / val_ws=76 at step 4443 — clearly diverging or coordinate-frequency mismatch. Don't waste 4+ more GPU-hours.

2. **Don't run sigma=2.0.** I'd rather have your sigma=1.0 win merged today than wait on a 3-arm sweep. If sigma=2.0 had won, you'd see it in val curves vs sigma=1.0 trajectory — and your own coord-scale analysis already calls out that the existing sincos `pos_embed` covers far-field volume coords. The sigma sweep is good but lower-priority than landing the win.

3. **Mark this PR ready for review.** I'll merge based on the sigma=1.0 result.

The sigma sweep is a great Round-2 follow-up if we want to push further (e.g., per-modality sigma, learnable B matrix, or much higher num_features). For now, your tay/DDP8 SOTA is already in.

**Comment by morganmcg1** [comment]
STUDENT fern:

## Results

Acting on advisor instructions: killed sigma=0.5 (pid tree at 15:32 UTC, ~step 4500/27210, val_abupt diverging at ~74) and skipped sigma=2.0. The detached orchestrator (pid 20078) is also dead, so no further arms will launch. Submitting sigma=1.0 as the final result.

### Headline (sigma=1.0, W&B `u43lik5d`)

| Metric | fern test | tay baseline | yi best | AB-UPT |
|---|---:|---:|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **17.77** | 19.81 | 15.82 | — |
| `test_primary/surface_pressure_rel_l2_pct` | **11.20** | 12.86 | 9.99 | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | **18.68** | 21.27 | 16.60 | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | **16.13** | 15.91 | 14.21 | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | **16.20** | 18.24 | 14.27 | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | **21.81** | 25.50 | 19.49 | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | **23.53** | 26.53 | 21.12 | 3.63 |

Best val (epoch 9): `val_primary/abupt_axis_mean_rel_l2_pct = 17.06`. Test is +0.72 above best val — typical generalization gap.

**Wins:** Every surface axis improves vs tay baseline. surface_pressure −1.66pp (−12.9%), wall_shear −2.59pp (−12.2%), tau axes −9–13%.
**No-op axis:** volume_pressure 16.13 vs 15.91 baseline (+0.22pp, parity within noise). Matches the coord-scale concern I flagged: at meter-scale `sigma=1.0`, far-field volume coords (~[-40, 80]m) saturate the RFF projection, so volume routes through the existing sincos `pos_embed` and gets no lift. Surface coords (≈[-1, 4]m) sit in the sigma=1.0 sweet spot.

### Run details

- **W&B group:** [fern-rff-sigma-sweep](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/groups/fern-rff-sigma-sweep)
- **Winning run:** [u43lik5d](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/u43lik5d) — sigma=1.0, 8×H100 DDP, `--no-compile-model`
- **Wall time:** 17,219 s ≈ 4.78 h (10 epochs to best, best_epoch=9)
- **Peak GPU mem:** ~74.5 GiB / 96 GiB across all 8 ranks
- **Trainable params:** 12,765,733 (delta vs no-RFF: +65,536 from `64 → 512` projection weight on each modality, as predicted)
- **Buffer params (fixed B matrix):** 192 elements (3×32 surface + 3×32 volume), non-trainable

### Reproduce

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --wandb-name "fern/rff-coord-sigma1.0" \
  --wandb-group "fern-rff-sigma-sweep" \
  --volume-loss-weight 2.0 \
  --batch-size 4 \
  --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --rff-num-features 32 --rff-sigma 1.0 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms \
  --no-compile-model
```

### What happened

RFF on the 3 coord channels gives the model a high-frequency basis for fine surface variation, exactly where it was spectrally bottlenecked by raw-coord input. Surface targets (pressure, wall shear vector) all benefit; volume targets don't, because at meter-scale sigma=1.0 the wider far-field volume coords overshoot the bandwidth. The existing sincos `pos_embed` covers volume already, so the projection learns to ignore noisy RFF channels there — explaining parity rather than regression.

Compile path stayed off (per advisor's compile-bug heads-up) — throughput dropped ~1.5–2× as expected, but 10 epochs in the budget were enough for clean curves and a clear win.

### Suggested follow-ups

1. **Per-modality sigma:** surface uses sigma≈1.0, volume uses sigma≈0.05–0.1 (matches its 80m range). Cleanest path to volume_pressure improvement without losing the surface lift.
2. **Coordinate normalization first:** scale volume coords to roughly the surface range before RFF, then a single sigma works everywhere. Cheaper than dual heads.
3. **Higher num_features (64 or 128):** trainable param delta still small (~256k for 128 features); may extract more from the fine-detail axes (tau_y/tau_z still 22–24).
4. **Learnable B:** initialise as Gaussian, set `requires_grad=True`. Cheap, unlikely to hurt, may pick up direction-dependent geometry priors.
5. **Round-2 idea:** combine RFF with the next-best round-1 winner so we stack rather than re-run sweeps.
